buildSrc/src/main/java/org/elasticsearch/gradle/tar/SymbolicLinkPreservingTar.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle.tar;
+
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream;
+import org.gradle.api.GradleException;
+import org.gradle.api.file.RegularFile;
+import org.gradle.api.internal.file.CopyActionProcessingStreamAction;
+import org.gradle.api.internal.file.archive.compression.ArchiveOutputStreamFactory;
+import org.gradle.api.internal.file.archive.compression.Bzip2Archiver;
+import org.gradle.api.internal.file.archive.compression.GzipArchiver;
+import org.gradle.api.internal.file.archive.compression.SimpleCompressor;
+import org.gradle.api.internal.file.copy.CopyAction;
+import org.gradle.api.internal.file.copy.CopyActionProcessingStream;
+import org.gradle.api.internal.file.copy.FileCopyDetailsInternal;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.WorkResult;
+import org.gradle.api.tasks.WorkResults;
+import org.gradle.api.tasks.bundling.AbstractArchiveTask;
+import org.gradle.api.tasks.bundling.Compression;
+import shadow.org.apache.tools.tar.TarConstants;
+import shadow.org.apache.tools.tar.TarOutputStream;
+import shadow.org.apache.tools.zip.UnixStat;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A custom archive task that assembles a tar archive that preserves symbolic links.
+ *
+ * This task is necessary because the built-in task {@link org.gradle.api.tasks.bundling.Tar} does not preserve symbolic links.
+ */
+public class SymbolicLinkPreservingTar extends AbstractArchiveTask {
```

Comment:
@mark-vieira I have mixed feelings about this. On the one hand, it feels not right because `SymbolicLinkPreservingTar` basically overrides all of the behavior of `Tar`, and looking at the `Tar` implementation, I don't think it was written intending to be overridden (if only it were that Java was `final` by default 🤦‍♀). On the other hand, I get your argument about `tasks.withType(Tar)` but I feel if we reach that point, we could as easily write that as `tasks.withType(SymbolicLinkPreservingTar)` 🤷‍♀). Extending merely to get reuse on `getCompression` isn't particularly compelling to me since it's straightforward either way, and we still have to reimplement the output stream compression logic anyway (again, from `Tar` not being designed for extensibility).

I don't feel strongly about this, but I do have a slight preference to not extend `Tar` in this case.

Reply:
The only thing we'd be overriding from `Tar` is `createCopyAction()` which is already an abstract method intended for subclasses to implement so I think in this very specific scenario is ok. Also, the implementation is for all intents and purposes backwards compatible with `Tar`, minus the symlink behavior which you could argue is really a "fix".

I think we are more likely to run into funkiness with these tasks in combination with logic we _expect_ to apply to all `Tar` tasks, than the other way around. The risk with custom task types is that it might break cross-cutting behaviors. Subclassing `Tar` mostly eliminates that problem.

Reply:
In any case, even `AbstractArchiveTask` isn't intended for subclassing outside the Gradle core codebase as it's not public API 😉.